File: 117_populating_next_right_pointers_in_each_node_II.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    private TreeNode prev = null;
    
    public void connect(TreeLinkNode root) {
        TreeLinkNode head = null; //head of the next level
        TreeLinkNode prev = null; //the leading node on the next level
        TreeLinkNode cur = root;  //current node of current level

        while (cur != null) {
            
            while (cur != null) { //iterate on the current level
                //left child
                if (cur.left != null) {
                    if (prev != null) {
                        // make connection to next right node
                        prev.next = cur.left;
                    } else {
                        head = cur.left;
                    }
                    prev = cur.left;
                }
                //right child
                if (cur.right != null) {
                    if (prev != null) {
                        // make connection to next right node
                        prev.next = cur.right;
                    } else {
                        head = cur.right;
                    }
                    // then jump to right node
                    prev = cur.right;
                }
                //move to next node
                // by using previous built connection to move curr to the right
                cur = cur.next;
            }
            
            //move to next level
            cur = head;
            head = null;
            prev = null;
        }
    }
}

File: python/117_populating_next_right_pointers_in_each_node_II.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left, right, next):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None

        # set up a dummy node at the most left of curr level
        # whenever linking next pointer for 'node'
        # we keep the previous link
        # that when node == None, we can keep track for previous layer
        curr = dummy = Node()
        node = root
        while node:
            if node.left:
                curr.next = node.left
                curr = node.left
            if node.right:
                curr.next = node.right
                curr = node.right
            node = node.next
            if not node:
                node = dummy.next
                dummy.next = None
                curr = dummy
        return root
